src/useSpring.ts

```diff
@@ -1,31 +1,37 @@
-import { useEffect, useState } from 'react';
+import { useEffect, useState, useMemo } from 'react';
 import { Spring, SpringSystem } from 'rebound';
 
 const useSpring = (targetValue: number = 0, tension: number = 50, friction: number = 3) => {
   const [spring, setSpring] = useState<Spring | null>(null);
   const [value, setValue] = useState<number>(targetValue);
 
-  useEffect(() => {
-    const listener = {
+  // memoize listener to being able to unsubscribe later properly, otherwise
+  // listener fn will be different on each re-render and wouldn't unsubscribe properly.
+  const listener = useMemo(
+    () => ({
       onSpringUpdate: currentSpring => {
         const newValue = currentSpring.getCurrentValue();
         setValue(newValue);
       },
-    };
+    }),
+    []
+  );
 
+  useEffect(() => {
     if (!spring) {
       const newSpring = new SpringSystem().createSpring(tension, friction);
       newSpring.setCurrentValue(targetValue);
       setSpring(newSpring);
       newSpring.addListener(listener);
-      return;
     }
 
     return () => {
-      spring.removeListener(listener);
-      setSpring(null);
+      if (spring) {
+        spring.removeListener(listener);
+        setSpring(null);
+      }
     };
-  }, [tension, friction]);
+  }, [tension, friction, spring]);
```

Comment:
Second issue: spring wasn't added to deps dependency so useEffect wasn't updated its callback to reference proper spring. What does that imply? Listener was never removed.

I usually work with vanilla js and I always set up this [ESLint plugin](https://www.npmjs.com/package/eslint-plugin-react-hooks) to get proper validation on hooks deps. This plugin would have detected this (or at least warned you) but sadly it's not available yet for TSLint: I found [this equivalent plugin](https://www.npmjs.com/package/tslint-react-hooks) but it doesn't implement the deps check yet (tho [it has an open issue for it](https://github.com/Gelio/tslint-react-hooks/issues/9), so I hope we can get it soon for TSLint)

Reply:
I think `react-use` has to switch to ESLint in foreseeable future, as TSLint is deprecated.